src/types/config.js

```diff
@@ -0,0 +1,95 @@
+ * particular function is provided.
+ *
+ * @typedef Service
+ * @prop {string} apiUrl
+ * @prop {string} authority
+ * @prop {string} grantToken
+ * @prop {boolean} [allowLeavingGroups]
+ * @prop {boolean} [enableShareLinks]
+ * @prop {Function} [onLoginRequest]
+ * @prop {boolean} [onLoginRequestProvided]
+ * @prop {Function} [onLogoutRequest]
+ * @prop {boolean} [onLogoutRequestProvided]
+ * @prop {Function} [onSignupRequest]
+ * @prop {boolean} [onSignupRequestProvided]
+ * @prop {Function} [onProfileRequest]
+ * @prop {boolean} [onProfileRequestProvided]
+ * @prop {Function} [onHelpRequest]
+ * @prop {boolean} [onHelpRequestProvided]
+ */
+
+/**
+ * Configuration for the Sentry crash-reporting service.
+ *
+ * @typedef SentryConfig
+ * @prop {string} dsn
+ * @prop {string} environment
+ */
+
+/**
+ * Configuration for the sidebar app set by the Hypothesis backend ("h")
+ * or baked into the sidebar app at build time (in the browser extension).
+ *
+ * See `h.views.client` in the "h" application.
+ *
+ * @typedef SidebarSettings
+ * @prop {string} apiUrl
+ * @prop {string} authDomain
+ * @prop {string} [googleAnalytics]
+ * @prop {string} oauthClientId
+ * @prop {string[]} rpcAllowedOrigins
+ * @prop {SentryConfig} [sentry]
+ * @prop {string} [websocketUrl]
+ */
+
+/**
+ * @typedef RequestConfigFromFrameOptions
+ * @prop {number} ancestorLevel
+ * @prop {string} origin
+ */
+
+/**
+ * Configuration set by the embedder of the client and used by the sidebar.
+ *
+ * This is the subset of keys from
+ * https://h.readthedocs.io/projects/client/en/latest/publishers/config/ which
+ * excludes any keys used only by the "annotator" part of the application.
+ *
+ * @typedef HostSettings
+ * @prop {string} [annotations] - Direct-linked annotation ID
+ * @prop {string} [group] - Direct-linked group ID
+ * @prop {string} [query] - Initial filter query
+ * @prop {string} [appType] - Method used to load the client
+ * @prop {boolean} [openSidebar] - Whether to open the sidebar on the initial load
+ * @prop {boolean} [showHighlights] - Whether to show highlights
+ * @prop {Object} [branding] -
+ *   Theme properties (fonts, colors etc.)
+ * @prop {boolean} [enableExperimentalNewNoteButton] -
+ *   Whether to show the "New note" button on the "Page Notes" tab
+ * @prop {RequestConfigFromFrameOptions|string} [requestConfigFromFrame]
+ *   Origin of the ancestor frame to request configuration from
+ * @prop {Service[]} [services] -
+ *   Configuration for the annotation services that the client connects to
+ * @prop {string} [theme]
+ *   Name of the base theme to use.
+ * @prop {string} [usernameUrl]
+ *   URL template for username links
+ */
+
+/**
+ * The `settings` object used in the sidebar app that is a result of merging
+ * (filtered and validated) configuration from the host page with configuration
+ * from h / the browser extension.
+ *
+ * @typedef {HostSettings & SidebarSettings} MergedSettings
```

Comment:
This is an example of creating an "intersection" type which combines all the properties from two existing types.